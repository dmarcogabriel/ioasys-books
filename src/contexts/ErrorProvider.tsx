import React, {createContext} from 'react';
import {UNAUTHORIZED_STATUS} from '../constants';
import Toast from 'react-native-toast-message';

type ErrorData = {
  message: string;
};

export interface ApiError {
  errors: ErrorData;
}

interface ErrorContextValue {
  handleError: (
    error: ApiError,
    status: number,
    logoutCallback: () => Promise<void>,
  ) => Promise<void>;
}

export const ErrorContext = createContext<ErrorContextValue>({
  handleError: async () => {},
});

interface ErrorProviderProps {
  children: React.ReactNode;
}

export const ErrorProvider = ({children}: ErrorProviderProps): JSX.Element => {
  const handleError = async (
    error: ApiError,
    status: number,
    logoutCallback: () => Promise<void>,
  ): Promise<void> => {
    let errorMessage: string = '';
    if (status === UNAUTHORIZED_STATUS) {
      await logoutCallback();
      errorMessage = 'Sess√£o expirada';
    } else {
      Toast.show({
        position: 'bottom',
        text1: errorMessage || error.errors.message,
        type: 'error',
      });
    }
  };

  return (
    <ErrorContext.Provider value={{handleError}}>
      {children}
    </ErrorContext.Provider>
  );
};
