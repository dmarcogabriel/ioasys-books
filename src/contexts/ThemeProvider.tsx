import React from 'react';
import {ThemeProvider as SCThemeProvider} from 'styled-components/native';

interface ThemeProviderProps {
  children: React.ReactNode;
}

export const ThemeProvider = ({children}: ThemeProviderProps): JSX.Element => {
  const theme = {
    colors: {
      primary: '#B22E6F',
      dark: '#333333',
      light: '#ffffff',
      grey: '#999999',
    },
  };

  return <SCThemeProvider theme={theme}>{children}</SCThemeProvider>;
};
