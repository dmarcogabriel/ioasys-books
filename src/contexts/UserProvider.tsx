import React, {createContext, useState, useEffect, useCallback} from 'react';
import {User} from '../@types/user.interface';
import {storeData, removeData, getData} from '../utils/storage';
import {AUTH_TOKEN_STORAGE_KEY} from '../constants';

interface UserContextValue {
  user: User | null;
  login: (user: User, token: string) => Promise<void>;
  logout: () => Promise<void>;
  token: string | null;
}

export const UserContext = createContext<UserContextValue>({
  user: null,
  token: null,
  login: async () => {},
  logout: async () => {},
});

interface UserProviderProps {
  children: React.ReactNode;
}

export const UserProvider = ({children}: UserProviderProps): JSX.Element => {
  const [user, setUser] = useState<User | null>(null);
  const [token, setToken] = useState<string | null>(null);
  const [isLoading, setIsLoading] = useState(true);

  const login = async (data: User, authToken: string): Promise<void> => {
    setUser(data);
    setToken(authToken);
    await storeData(AUTH_TOKEN_STORAGE_KEY, authToken);
  };

  const logout = async () => {
    await removeData(AUTH_TOKEN_STORAGE_KEY);
    setUser(null);
    setToken(null);
  };

  const loadToken = useCallback(async () => {
    setIsLoading(true);
    try {
      const authToken = await getData<string>(AUTH_TOKEN_STORAGE_KEY);
      if (authToken) {
        setToken(authToken);
      }
    } catch (error) {
      setIsLoading(false);
    }
    setIsLoading(false);
  }, []);

  useEffect(() => {
    loadToken();
  }, [loadToken]);

  return (
    <UserContext.Provider value={{user, token, login, logout}}>
      {!isLoading && children}
    </UserContext.Provider>
  );
};
