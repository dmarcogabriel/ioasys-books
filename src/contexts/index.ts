export * from './UserProvider';
export * from './ThemeProvider';
export * from './ErrorProvider';
