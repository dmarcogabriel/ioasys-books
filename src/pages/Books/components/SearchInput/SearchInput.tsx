import React, {useState} from 'react';
import {
  SearchInputContainer,
  SearchInputStyled,
  SearchInputButton,
} from './styles';
import MIcon from 'react-native-vector-icons/MaterialIcons';

interface SearchInputProps {
  onSearch: (term: string) => void;
}

export const SearchInput = ({onSearch}: SearchInputProps): JSX.Element => {
  const [value, setValue] = useState<string>('');

  const handleChange = (term: string) => setValue(term);

  const handleSearch = () => onSearch(value);

  return (
    <SearchInputContainer>
      <SearchInputStyled
        placeholder="Procure um livro"
        value={value}
        onChangeText={handleChange}
        onEndEditing={handleSearch}
      />
      <SearchInputButton onPress={handleSearch}>
        <MIcon name="search" size={24} />
      </SearchInputButton>
    </SearchInputContainer>
  );
};
