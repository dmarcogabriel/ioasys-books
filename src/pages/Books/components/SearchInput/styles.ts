import styled from 'styled-components/native';

export const SearchInputContainer = styled.View`
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  border: 1px solid #33333333;
  width: 90%;
  border-radius: 2px;
`;

export const SearchInputStyled = styled.TextInput`
  padding: 16px 46px 16px 8px;
  width: 100%;
`;

export const SearchInputButton = styled.TouchableOpacity`
  margin: 16px;
  position: absolute;
  right: 0;
`;
