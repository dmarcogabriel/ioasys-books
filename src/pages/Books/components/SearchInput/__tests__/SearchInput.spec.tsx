import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import {SearchInput, SearchInputProps} from '../SearchInput';
import {ThemeProvider} from '../../../../../contexts';
import renderer from 'react-test-renderer';

const ComponentWrapper = (props: SearchInputProps) => (
  <ThemeProvider>
    <SearchInput {...props} />
  </ThemeProvider>
);
let mockOnSearchResult: string | null = null;
const mockOnSearch = jest.fn((term: string) => {
  mockOnSearchResult = term;
});

describe('Books/components/SearchInput', () => {
  it('should pass on change input', () => {
    const {getByTestId} = render(<ComponentWrapper onSearch={mockOnSearch} />);
    const searchInput = getByTestId('searchInput');

    fireEvent.changeText(searchInput, 'Test');

    expect(searchInput).toHaveProp('value', 'Test');
  });

  it('should pass on submit search', () => {
    const {getByTestId} = render(<ComponentWrapper onSearch={mockOnSearch} />);
    const searchInput = getByTestId('searchInput');
    const searchInputButton = getByTestId('searchInputButton');

    fireEvent.changeText(searchInput, 'Test 2');
    fireEvent.press(searchInputButton);
    expect(mockOnSearch).toHaveBeenCalled();
    expect(mockOnSearchResult).toBe('Test 2');
  });

  it('should match snapshot', () => {
    expect(
      renderer.create(<ComponentWrapper onSearch={mockOnSearch} />).toJSON(),
    ).toMatchSnapshot();
  });
});
