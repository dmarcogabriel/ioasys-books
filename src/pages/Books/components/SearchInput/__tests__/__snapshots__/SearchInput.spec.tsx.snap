// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Books/components/SearchInput should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderColor": "#33333333",
        "borderRadius": 2,
        "borderStyle": "solid",
        "borderWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": "90%",
      },
    ]
  }
>
  <TextInput
    onChangeText={[Function]}
    onEndEditing={[Function]}
    placeholder="Procure um livro"
    style={
      Array [
        Object {
          "paddingBottom": 16,
          "paddingLeft": 8,
          "paddingRight": 46,
          "paddingTop": 16,
          "width": "100%",
        },
      ]
    }
    testID="searchInput"
    value=""
  />
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "marginBottom": 16,
        "marginLeft": 16,
        "marginRight": 16,
        "marginTop": 16,
        "opacity": 1,
        "position": "absolute",
        "right": 0,
      }
    }
    testID="searchInputButton"
  >
    <Text>
      Icon
    </Text>
  </View>
</View>
`;
